app/src/utilities.lisp

```diff
@@ -111,16 +116,18 @@ string followed by optional args (as in FORMAT)."
     ;; time.
     (cl-syslog:get-priority level-or-fmt-string))
   (if (keywordp level-or-fmt-string)
-      `(cl-syslog:format-log
-        *logger*
-        ',level-or-fmt-string
-        "~A~@?"
-        (session-info)
-        ,@fmt-string-or-args)
-      `(cl-syslog:format-log
-        *logger*
-        ':debug
-        "~A~@?"
-        (session-info)
-        ,level-or-fmt-string
-        ,@fmt-string-or-args)))
+      `(with-locked-log ()
```

Comment:
Can `cl-syslog` handle multithreaded logging with a separate logger per thread? Or is it just broken in general?  I guess as long as the lock is not heavily contended: no big deal. But if the QVM is logging `:debug` messages like a banshee somewhere, that would be sad. Infinitely better than segfaulting, of course.

Reply:
I'm not sure it can; it all boils down to calling `syslog(3)` which I'm not sure is re-entrant.

Reply:
I think POSIX mandates thread-safety for `syslog`? Doesn't mean it's implemented that way everywhere, but I'd be mildly surprised if it weren't the case on Linux at least.

http://pubs.opengroup.org/onlinepubs/000095399/functions/xsh_chap02_09.html#tag_02_09_01

Reply:
"man syslog" on my Linux latop claims:

```
ATTRIBUTES
       For an explanation of the terms used in this section, see attributes(7).

       ┌──────────────────────┬───────────────┬────────────────────┐
       │Interface             │ Attribute     │ Value              │
       ├──────────────────────┼───────────────┼────────────────────┤
       │openlog(), closelog() │ Thread safety │ MT-Safe            │
       ├──────────────────────┼───────────────┼────────────────────┤
       │syslog(), vsyslog()   │ Thread safety │ MT-Safe env locale │
       └──────────────────────┴───────────────┴────────────────────┘
```


`man 7 attributes` says:

```
MT-Safe

MT-Safe or Thread-Safe functions are safe to call in the presence of
other threads.  MT, in MT-Safe, stands for Multi Thread.

Being MT-Safe does not imply a function is atomic, nor that it uses
any of the memory synchronization mechanisms POSIX exposes to users.

It is even possible that calling MT-Safe functions in sequence does
not yield an MT-Safe combination.  For example, having a thread call
two MT-Safe functions one right after the other does not guarantee
behavior equivalent to atomic execution of a combination of both
functions, since concurrent calls in other threads may interfere in a
destructive way.
...
```

so maybe `MT-Safe` doesn't mean your concurrent `syslog` calls won't trample on each other and result in garbled logs, just that they shouldn't segfault.  I guess this is a "suck it and see" scenario!

Reply:
These attributes don't make sense to me. (Also, I should say I only tested on macOS.)

Reply:
Indeed, what exactly is implied by MT-Safe is not perfectly clear to me. I also think I'm conflating two separate concepts, namely "thread safe" vs "re-entrant".

In any case, I wrote a simple test and it appears to have "worked on my machine", but perhaps the test is too simple or not doing what I think it is.

Here is the test:

```lisp
(ql:quickload :lparallel)
(ql:quickload :cl-syslog)

(defun make-logger (n &optional tee)
  (let* ((app-name (format nil "cl-syslog-test-~D" n))
         (log-writer (cl-syslog:syslog-log-writer app-name :local0)))
    (make-instance
     'cl-syslog:rfc5424-logger
     :app-name app-name
     :facility ':local0
     :maximum-priority ':debug
     :log-writer
     (or (and tee (cl-syslog:tee-to-stream log-writer *error-output*))
         log-writer))))

(setf lparallel:*kernel* (lparallel:make-kernel 4))

(let ((sleep t)
      (ch (lparallel:make-channel))
      (num-tasks (lparallel:kernel-worker-count)))
  (dotimes (n num-tasks)
    (lparallel:submit-task
     ch
     (let ((logger (make-logger n nil))
           (message (coerce (loop :repeat 20 :collect (code-char (+ n 65)))
                            'string)))
       (lambda ()
         (dotimes (i 1000)
           (cl-syslog:format-log logger ':debug message)
           (when (and sleep (zerop (random 4)))
             (sleep (random 1.0))))))))
  (loop :repeat num-tasks :collect (lparallel:receive-result ch)))

(sb-ext:quit)
```

I tried running it both in slime and on the command line via `sbcl --load ...`, with and without the `sleep`s, with and without teeing logs to `*error-output*`, etc. In every case I got reasonable looking logs in my `journalctl` and on stderr. I see log lines interleaved, but never trampling on each other. I'm assuming `lparallel` is really running the tasks in parallel, and not just round-robin'ing among them. I guess it's also possible those blocking syslog calls are slowing things down enough to avoid havoc... but then I suppose that would also be the case in the QVM.